datafusion/expr/src/expr.rs

```diff
@@ -983,10 +983,290 @@ impl PartialOrd for Expr {
 }
 
 impl Expr {
-    /// Returns the name of this expression as it should appear in a schema. This name
-    /// will not include any CAST expressions.
-    pub fn display_name(&self) -> Result<String> {
-        create_name(self)
+    /// Returns the name for schema / field that is different from Display
+    /// Most of the expressions are the same as Display
+    /// Those are the main difference
+    /// 1. Alias, where excludes expression
+    /// 2. Cast / TryCast, where takes expression only
+    pub fn schema_name(&self) -> Result<String> {
```

Comment:
If we are going to introduce a new API anyways, what do you think about making an API that doesn't *require* allocating a new string and making it infallable. Something like

```suggestion
    pub fn schema_name<'a>(&'a self) -> impl Display + 'a {
```

We could probably then implement a new type thing and mostly use the same code:

```rust
// new type wrapper
struct SchemaDisplay<'a>(&'a Expr);

impl <'a> Display for SchemaDisplay<'a> {
  fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
        match self {
            // The same as Display
            Expr::Column(_)
            | Expr::Literal(_)
            | Expr::ScalarVariable(..)
            | Expr::Sort(_)
            | Expr::OuterReferenceColumn(..)
            | Expr::Placeholder(_)
            | Expr::Wildcard { .. } => write!(f, "{self}")?,
...
  }
}

return SchemaDisplay(self)
```


The benefit of doing this would be anywhere it needed to get formatted would not require a copy

```rust
println!("This would not allocated a String: {}", expr.schema_name());
```

```rust
// get schema name as a string:
let schema_name = expr.schema_name().to_string();

Reply:
This looks cool!